cpp/src/io/json/json_tree.cu

```diff
@@ -439,210 +457,227 @@ rmm::device_uvector<size_type> hash_node_type_with_field_name(device_span<Symbol
+                        stream.value()};
+  // path compression is not used since extra writes make all map operations slow.
+  auto d_hasher = [node_level      = node_levels.begin(),
+                   node_type       = node_type.begin(),
+                   parent_node_ids = parent_node_ids.begin()] __device__(auto node_id) {
+    auto hash =
+      cudf::detail::hash_combine(cudf::detail::default_hash<TreeDepthT>{}(node_level[node_id]),
+                                 cudf::detail::default_hash<size_type>{}(node_type[node_id]));
+    node_id = parent_node_ids[node_id];
+    while (node_id != parent_node_sentinel) {
+      hash = cudf::detail::hash_combine(
+        hash, cudf::detail::default_hash<TreeDepthT>{}(node_level[node_id]));
+      hash = cudf::detail::hash_combine(
+        hash, cudf::detail::default_hash<size_type>{}(node_type[node_id]));
+      node_id = parent_node_ids[node_id];
+    }
+    return hash;
+  };
+
+  rmm::device_uvector<hash_value_type> node_hash(num_nodes, stream);
+  thrust::tabulate(rmm::exec_policy(stream), node_hash.begin(), node_hash.end(), d_hasher);
+  auto d_hashed_cache = [node_hash = node_hash.begin()] __device__(auto node_id) {
+    return node_hash[node_id];
+  };
+
+  auto d_equal = [node_level      = node_levels.begin(),
+                  node_type       = node_type.begin(),
+                  parent_node_ids = parent_node_ids.begin(),
+                  d_hashed_cache] __device__(auto node_id1, auto node_id2) {
+    if (node_id1 == node_id2) return true;
+    if (d_hashed_cache(node_id1) != d_hashed_cache(node_id2)) return false;
+    auto is_equal_level = [node_level, node_type](auto node_id1, auto node_id2) {
+      if (node_id1 == node_id2) return true;
+      return node_level[node_id1] == node_level[node_id2] and
+             node_type[node_id1] == node_type[node_id2];
+    };
+    // if both nodes have same node types at all levels, it will check until it has common parent
+    // or root.
+    while (node_id1 != parent_node_sentinel and node_id2 != parent_node_sentinel and
+           node_id1 != node_id2 and is_equal_level(node_id1, node_id2)) {
+      node_id1 = parent_node_ids[node_id1];
+      node_id2 = parent_node_ids[node_id2];
+    }
+    return node_id1 == node_id2;
+  };
+
+  // insert and convert node ids to unique set ids
+  auto num_inserted = thrust::count_if(
+    rmm::exec_policy(stream),
+    thrust::make_counting_iterator<size_type>(0),
+    thrust::make_counting_iterator<size_type>(num_nodes),
+    [d_hashed_cache,
+     d_equal,
+     view       = key_map.get_device_mutable_view(),
+     uq_node_id = col_id.begin()] __device__(auto node_id) mutable {
+      // typename hash_map_type::value_type const insert_pair{};
+      auto it = view.insert_and_find(cuco::make_pair(node_id, node_id), d_hashed_cache, d_equal);
+      uq_node_id[node_id] = (it.first)->first;  // first.load(cuda::std::memory_order_relaxed);
+      return it.second;
+    });
```

Comment:
This seems like it could be implicitly relying on count_if calling the lambda only once. Would any of the bulk update functions of key_map be suitable? Also, as a future improvement, since you already precompute the hash values, you could maybe also sort them and improve memory access patterns/collision resolution that way?

Reply:
Bulk update functions does not return inserted keys for each key. 
Also,`insert_and_find` does not update size of the map (@PointKernel). 
That's why we depend on count_if. But this is a valid concern to relying on `count_if` on calling the lambda only once for each input iterator element. `count_if`'s internal implentation may be bound to change (cub::deviceReduce).

Adding a custom reduce kernel for this could avoid dependency on these library implementations. what do you think? @upsj 

Sorting is a heavy operation. (in terms of both memory and compute). I will try it in another PR.

Reply:
> insert_and_find does not update size of the map

That's correct. We don't update `size` counter in device view insertion. The use of `static_map` looks good to me. This can be further optimized once `cuco::static_set` is ready.